### DISCUSSION TITLE
Why is "Smart Charge" a switch and not a sensor?

### DISCUSSION THREAD

**Discussion by HLFCode** [BottlecapDave/HomeAssistant-OctopusEnergy#1004] [Q&A] [answered]
I thought smart charging is controlled by OE so what does changing the state of [Smart Charge](https://bottlecapdave.github.io/HomeAssistant-OctopusEnergy/entities/intelligent/#smart-charge) in HA actually do?
It doesn't seem to change state even when OE fires up my charger during the day
(I'm very reluctant to just try it!)

Excellent integration :-)

**Comment by BottlecapDave** [comment, marked as answer]
The smart charge switch turns off OE controlling your smart charger/car and stops it producing dispatch periods. It's the same as the switch specified in this forum within the app https://myenergi.info/viewtopic.php?p=116829#p116829

**Comment by HLFCode** [comment]
Thanks for the quick answer @BottlecapDave.
I've found it on the app too (well hidden).
Very useful as I no longer have to unplug the car if "tomorrow" is going to be sunny.

**Comment by HLFCode** [comment]
I think OE has disconnected the **Smart Charge** switch.
Changing the OE app setting is not reflected in the HA switch (always off).
Changing the HA switch to on is cancelled a few minutes later (presumably on the next update with OE) irrespective of the OE app setting.

Anybody else seeing this?

**Reply by BottlecapDave** [reply]
It might be something that is not supported by the provider or something has changed in the latest API changes (if you're using the latest beta). If you feel comfortable, could you execute the following graphql query following the instructions in https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/discussions/878 both when you have flipped the switch on and off in the app? I'm not on intelligent so rely on the community for areas like this.

```
query {
  devices(accountNumber: "A-XXX") {
    id
    deviceType
    provider
    status {
      currentState
      isSuspended
    }
    ... on SmartFlexVehicle {
      make
      model
      chargingPreferences {
        weekdayTargetTime
        weekdayTargetSoc
        weekendTargetTime
        weekendTargetSoc
        minimumSoc
        maximumSoc
      }
    }
    ... on SmartFlexChargePoint {
      make
      model
      chargingPreferences {
        weekdayTargetTime
        weekdayTargetSoc
        weekendTargetTime
        weekendTargetSoc
        minimumSoc
        maximumSoc
      }
    }
  }
}
```

**Comment by HLFCode** [comment]
Took me a while to grasp the GraphQL interface
With the car unplugged:

When "On" in app
```
{
  "data": {
    "devices": [
      {
        "id": "xxxxxxxxxxxxxxxxxxxxxxxxxxxxx",
        "deviceType": "ELECTRIC_VEHICLES",
        "provider": "MYENERGI",
        "status": {
          "currentState": "SMART_CONTROL_NOT_AVAILABLE",
          "isSuspended": false
        },
        "make": "Myenergi",
        "model": "Zappi smart EV",
        "chargingPreferences": {
          "weekdayTargetTime": "09:00:00",
          "weekdayTargetSoc": 20,
          "weekendTargetTime": "09:00:00",
          "weekendTargetSoc": 20,
          "minimumSoc": null,
          "maximumSoc": null
        }
      }
    ]
  }
}

```
When "Off" in app
```
{
  "data": {
    "devices": [
      {
        "id": "xxxxxxxxxxxxxxxxxxxxxxxx",
        "deviceType": "ELECTRIC_VEHICLES",
        "provider": "MYENERGI",
        "status": {
          "currentState": "SMART_CONTROL_OFF",
          "isSuspended": true
        },
        "make": "Myenergi",
        "model": "Zappi smart EV",
        "chargingPreferences": {
          "weekdayTargetTime": "09:00:00",
          "weekdayTargetSoc": 20,
          "weekendTargetTime": "09:00:00",
          "weekendTargetSoc": 20,
          "minimumSoc": null,
          "maximumSoc": null
        }
      }
    ]
  }
}
```

Does this help?

For others I ended up at [Octopus' GraphQL server](https://api.octopus.energy/v1/graphql/)
I entered the get token mutation on the left pane where the api key if the key used in the HA integration config
```
mutation {
	obtainKrakenToken(input: { APIKey: "API_KEY" }) {
		token
	}
}

```
I then pasted the above query in the top left pane the returned token into the header pane (lower left)
```
{
    "Authorization": "12345"
}

```

**Comment by HLFCode** [comment]
With the car plugged in but "Off" in the app
```
{
  "data": {
    "devices": [
      {
        "id": "xxxxxxxxxxxxxxxxxxxxxxxxxxxx",
        "deviceType": "ELECTRIC_VEHICLES",
        "provider": "MYENERGI",
        "status": {
          "currentState": "SMART_CONTROL_OFF",
          "isSuspended": true
        },
        "make": "Myenergi",
        "model": "Zappi smart EV",
        "chargingPreferences": {
          "weekdayTargetTime": "09:00:00",
          "weekdayTargetSoc": 20,
          "weekendTargetTime": "09:00:00",
          "weekendTargetSoc": 20,
          "minimumSoc": null,
          "maximumSoc": null
        }
      }
    ]
  }
}

```
plugged in and "On" in the app

```
{
  "data": {
    "devices": [
      {
        "id": "xxxxxxxxxxxxxxx",
        "deviceType": "ELECTRIC_VEHICLES",
        "provider": "MYENERGI",
        "status": {
          "currentState": "SMART_CONTROL_CAPABLE",
          "isSuspended": false
        },
        "make": "Myenergi",
        "model": "Zappi smart EV",
        "chargingPreferences": {
          "weekdayTargetTime": "09:00:00",
          "weekdayTargetSoc": 20,
          "weekendTargetTime": "09:00:00",
          "weekendTargetSoc": 20,
          "minimumSoc": null,
          "maximumSoc": null
        }
      }
    ]
  }
}
```

**Reply by BottlecapDave** [reply]
The data coming back looks right. I'd suggest trying the latest beta release, which updates the intelligent sensors to use new APIs and see if this fixes your issues. The sensor will take up to 5 minutes to reflect the data in the app (if done via the app).

**Reply by HLFCode** [reply]
After a bit more testing I think I see what's happening here.

The OE app toggle (in accounts->devices) toggles the OE app devices status message "smart control off"/"smart control ready" when the car is plugged in and "smart control off"/"unplugged" when the car is not plugged in.
All as you would expect.

The HA smart charge switch also toggles the OE app devices status message as above.

The HA switch seems to have no effect on the OE setings toggle **unless I restart the app** so it looks like the app doesn't expect anybody else is capable of changing the state and so only reads it **in the settings** on startup.

However, the HA switch always resets to off after the next update irrespective of the OE app "smart control" switch and irrespective of the car being plugged in or not.

It looks like the HA switch is working as designed except it doesn't show the current state of OE's status even using 13.0.0-beta.2.

From the messages above it looks like the current state has one OFF status (SMART_CONTROL_OFF) and two on statuses (SMART_CONTROL_CAPABLE and SMART_CONTROL_NOT_AVAILABLE).
This means the "on" status has two values depending on whether or not the car is plugged in.

Currently, to turn the OE smart charge to off you need to turn the HA switch on then off. 
Also HA is unable to determine if it's on or off (for me anyway).

Hope this helps.

**Reply by BottlecapDave** [reply]
Thanks for the debugging. I'm currently using the is suspended field to determine if it should be on/off so sounds like I should move to the state instead. I'll try and do this as part of the next release.

**Comment by HLFCode** [comment]
>so sounds like I should move to the state instead

No, I'd missed isSuspended - it looks an easier option to stick with it as the values look right

I've had a look at the code and I think the error is [here](https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/blob/ffb8911dd0e6189f1573c5e5677b4a8b238f64af/custom_components/octopus_energy/api_client/__init__.py#L1089)

I think line 1089 is looking for a string instead of a boolean?
Should
` 
smart_charge = devices[0]["status"]["isSuspended"] == 'false' if "status" in devices[0] and "isSuspended" in devices[0]["status"] else None
`
be changed to
` 
smart_charge = devices[0]["status"]["isSuspended"] if "status" in devices[0] and "isSuspended" in devices[0]["status"] else None
`

**Reply by HLFCode** [reply]
Sorry, needs the inverse I think
`smart_charge = not devices[0]["status"]["isSuspended"] if "status" in devices[0] and "isSuspended" in devices[0]["status"] else None
`
or 
`smart_charge = devices[0]["status"]["isSuspended"] == False if "status" in devices[0] and "isSuspended" in devices[0]["status"] else None
`

**Reply by HLFCode** [reply]
Tried this on my HA and it now works as expected
`
smart_charge = devices[0]["status"]["isSuspended"] == False if "status" in devices[0] and "isSuspended" in devices[0]["status"] else None 
`

`== False` probably better than the `not xxx` version...

**Reply by BottlecapDave** [reply]
Thanks for testing the fix. I'll get it out in the next release
